[RequirementMachine] Diagnose invalid requirements detected before rule construction.

hborla opened the pull request:
This change diagnoses the following invalid requirements that are detected (and dropped) before rule construction:
* A conformance requirement to a non-protocol, non-class type, e.g. `T: Int`.
* Concrete same-type mismatches, e.g. `String == Int`. This is diagnosed as a conflict.
* Conformance/superclass/layout constraints on concrete types, e.g. `Int: P` and `[T]: Collection`. This is diagnosed either as a redundancy or as `type 'X' in requirement does not refer to a generic parameter or associated type`.
* Redundant same-type requirements, e.g. `Int == Int` or `Array<T> == Array<Int>`.

Requirement errors are represented with the `RequirementError` struct, which stores an error kind, a requirement, and a source location. Errors are recorded while computing a generic signature, and diagnosed in `diagnoseRequirementErrors`.

Currently, errors recorded with an invalid source location are simply skipped. Once source locations are plumbed through the rewrite system, a valid source location should become an assertion.

Resolves: rdar://88135553

hborla committed:
[RequirementMachine] Diagnose non-protocol, non-class conformance/subtype
constraints in the requirement machine.

hborla committed:
[RequirementMachine] Plumb the requirement errors vector through desugarRequirement.

hborla committed:
[RequirementMachine] Diagnose concrete same-type mismatches, such as
`where String == Int`.

hborla committed:
[RequirementMachine] Diagnose conformance, layout, and superclass constraints
on non-type-parameters in the requirement machine.

hborla committed:
[RequirementMachine] Diagnose early redundant requirements.

hborla committed:
[RequirementMachine] Always store an invalid requirement in RequirementError.

slavapestov commented on lib/AST/RequirementMachine/RequirementLowering.cpp:
```diff
@@ -108,9 +146,20 @@ static void desugarSuperclassRequirement(Type subjectType,
-    // FIXME: Diagnose redundancy or conflict upstream
+    Requirement requirement(RequirementKind::Layout,
+                            subjectType, layout);
+    if (subjectType->isAnyClassReferenceType()) {
```
This should be conditional on the requirement itself being a class layout requirement. You can't write any other kind of layout requirement in a normal generic signature, but you can in a `@_specialize`.

hborla replied:
Oh neat! Is there a way to check if a given type satisfies a given layout requirement, or should I just add the condition that the layout constraint is a class layout like you suggest?

slavapestov replied:
Unfortunately we don't have a way to statically check other layout requirements

slavapestov commented on include/swift/AST/Requirement.h:
```diff
@@ -124,6 +124,55 @@ struct StructuralRequirement {
+    : kind(kind), requirement(requirement), loc(loc) {}
+
+public:
+  static RequirementError forInvalidConformance(Type subjectType,
```
Perhaps forInvalidTypeRequirement() is a better name -- in this case it's not a conformance *or* a superclass requirement, it's just something like 'T : Int' right?

slavapestov commented on include/swift/AST/Requirement.h:
```diff
@@ -124,6 +124,55 @@ struct StructuralRequirement {
+    return {Kind::ConcreteTypeMismatch, requirement, loc};
+  }
+
+  static RequirementError forNonTypeParameter(Requirement req,
```
forConflictingRequirement() perhaps, since in reality this means the concrete type doesn't satisfy the requirement

slavapestov commented on lib/AST/RequirementMachine/RequirementLowering.cpp:
```diff
@@ -455,29 +526,136 @@ void swift::rewriting::realizeInheritedRequirements(
+    }
+
+    case RequirementError::Kind::NonTypeParameter: {
+      ctx.Diags.diagnose(loc, diag::requires_not_suitable_archetype,
```
Do we filter out the ErrorType case elsewhere? I imagine it's something like 'DoesNotExist : Int'

slavapestov commented on lib/AST/RequirementMachine/RequirementLowering.cpp:
```diff
@@ -455,29 +526,136 @@ void swift::rewriting::realizeInheritedRequirements(
   assert(!proto->hasLazyRequirementSignature());
 
   SmallVector<StructuralRequirement, 4> result;
+  SmallVector<RequirementError, 4> errors;
```
You can diagnose these here too -- this request runs for protocols defined in source

slavapestov commented on lib/AST/RequirementMachine/RequirementLowering.cpp:
```diff
@@ -556,6 +734,7 @@ TypeAliasRequirementsRequest::evaluate(Evaluator &evaluator,
   assert(!proto->hasLazyRequirementSignature());
 
   SmallVector<Requirement, 2> result;
+  SmallVector<RequirementError, 2> errors;
```
You can diagnose these here too -- this request runs for protocols defined in source

hborla committed:
[RequirementMachine] Emit diagnostics for structural requirements in protocol
and typealias requirement signatures.

slavapestov commented on test/Generics/requirement_machine_diagnostics.swift:
```diff
@@ -0,0 +1,65 @@
+// FIXME: Error emitted twice.
+func badTypeConformance5<T>(_: T) where T & Sequence : EqualComparable { }
+// expected-error@-1 2 {{non-protocol, non-class type 'T' cannot be used within a protocol-constrained type}}
+// expected-error@-2{{type 'Sequence' in conformance requirement does not refer to a generic parameter or associated type}}
```
Heh, I think I actually fixed this in https://github.com/apple/swift/pull/41381/commits/4655fc359be4cf5e09f28e1b3250649d4d269ee3

slavapestov commented on lib/AST/RequirementMachine/RequirementLowering.cpp:
```diff
@@ -455,29 +526,136 @@ void swift::rewriting::realizeInheritedRequirements(
   }
 }
 
+bool swift::rewriting::diagnoseRequirementErrors(
```
Can you add a documentation comment explaining the boolean return value?

slavapestov commented on include/swift/AST/Requirement.h:
```diff
@@ -124,6 +124,55 @@ struct StructuralRequirement {
   bool inferred = false;
 };
 
+/// Represents an invalid requirement, such as `T: Int`.
```
Do you think this will be used outside of lib/AST/RequirementMachine? If not, we could put this in lib/AST/RequirementMachine/Diagnostics.h or something.

slavapestov reviewed (commented):
This is great, I have a few minor comments

slavapestov commented:
> Once source locations are plumbed through the rewrite system, a valid source location should become an assertion.

FWIW we don't always have a source location, if the requirement came from another generic signature (even in the same file). Eg if you have

```struct Foo<T : P> {}

extension Foo where T == ConcreteType```

Then the T : P in struct Foo becomes redundant, but we don't diagnose it because there's no source location.

slavapestov commented on lib/AST/RequirementMachine/RequirementLowering.h:
```diff
@@ -40,18 +40,26 @@ namespace rewriting {
+
+bool diagnoseRequirementErrors(ASTContext &ctx,
+                               SmallVectorImpl<RequirementError> &errors,
+                               bool allowConcreteGenericParams);
```
BTW I think we want to diagnose allowConcreteGenericParams after the rewrite system is built, so maybe this parameter doesn't have to be here at all?

hborla replied:
`allowConcreteGenericParams` currently used in this function to decide whether to offer a fix-it to turn requirements like `T: Int` into `T == Int`

slavapestov replied:
Oh oops! Worth mentioning in a doc comment at the top of the function ;-)

hborla replied:
Done!

hborla committed:
[NFC][RequirementMachine] Add documentation for diagnoseRequirementErrors.

slavapestov commented on lib/AST/RequirementMachine/RequirementLowering.cpp:
```diff
@@ -455,29 +526,149 @@ void swift::rewriting::realizeInheritedRequirements(
+    ASTContext &ctx, SmallVectorImpl<RequirementError> &errors,
+    bool allowConcreteGenericParams) {
+  bool diagnosedError = false;
+  if (ctx.LangOpts.RequirementMachineProtocolSignatures !=
```
I think this check should happen in the callers, since it depends on RequirementMachineInferredSignatures also?

hborla replied:
Oh, I think you're right that this is wrong 🙂 it should be that `InferredGenericSignatureRequestRQM` checks `RequirementMachineInferredSignatures` and the other requests for protocol and typealias requirements check `RequirementMachineProtocolSignatures`, right?

slavapestov commented on test/Generics/requirement_machine_diagnostics.swift:
```diff
@@ -0,0 +1,73 @@
+// expected-warning@-1{{redundant conformance constraint 'S' : 'P'}}
+
+class C {}
+func concreteLayoutRedundancy<T>(_: T) where C : AnyObject {}
```
I think you need to test the layout conflict case (struct S / S : AnyObject)

slavapestov commented on test/Generics/requirement_machine_diagnostics.swift:
```diff
@@ -0,0 +1,73 @@
+
+class C2: C {}
+func concreteSubclassRedundancy<T>(_: T) where C2 : C {}
+// expected-warning@-1{{redundant superclass constraint 'C2' : 'C'}}
```
I think you need to test the superclass conflict case (class C / class D / C : D)

slavapestov commented on test/Generics/requirement_machine_diagnostics.swift:
```diff
@@ -0,0 +1,73 @@
+func badTypeConformance6<T>(_: T) where [T] : Collection { }
+// expected-warning@-1{{redundant conformance constraint '[T]' : 'Collection'}}
+
+func concreteSameTypeRedundancy<T>(_: T) where Int == Int {}
```
Same-type conflict case: Int == String

hborla replied:
This is tested earlier in the file in a protocol extension

```swift
// expected-error@+1{{generic signature requires types 'Double' and 'Int' to be the same}}
extension X where NotAnInt == Int {}
```

do you also want me to test this case in a function generic signature?

slavapestov replied:
Oh I missed that, oops! Well spotting 2 out of 3 missing tests isn't too bad.

hborla committed:
[NFC][RequirementMachine] Rename some of the RequirementError::Kind cases,
add documentation for each case, and move RequirementError to
lib/AST/RequirementMachine/Diagnostics.h.

hborla committed:
[RequirementMachine] Only emit a redundancy warning for layout constraints
if the type is constrained to a class layout, and add a few more test cases
for type requirement conflicts.

hborla committed:
[RequirementMachine] Suppress requirement diagnostics when one of the
types already has an error.

hborla committed:
[RequirementMachine] Move gating emitting diagnostics on the requirement
macine flags to callers.

Callers gate this decision on different flags; InferredGenericSignatureRequestRQM
needs RequirementMachineInferredSignatures to be enabled, and the other
callsites need RequirementMachineProtocolSignatures to be enabled.

slavapestov reviewed (approved):
Excellent, it looks ready to land to me.

hborla commented:
Thanks so much for the in-depth review!

hborla commented:
@swift-ci please smoke test

hborla committed:
[NFC][RequirementMachine] Fix duplicate diagnostics in requirement machine
tests.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test Linux
